addons/udes_common/tests/test_selection_display_name.py

```diff
@@ -0,0 +1,117 @@
+                user_id=cls.env.user.id,
+                model_id="res.groups",
+            )
+        )
+
+    def test_basic_selection_field(self):
+        """Check function gives the expected value from a normal selection field"""
+        normal_output = self.model_overview_report.report_type
+        self.assertEqual(normal_output, "qweb-pdf")
+
+        alt_output = self.model_overview_report.selection_value_displayname("report_type")
+        self.assertEqual(alt_output, "PDF")
+
+    def test_generated_selection_field(self):
+        """Check function gives the expected value from a selection field
+        which uses a function to determine its selection values
+        (where the function is defined in the field as a callable that lives outside the class)
+        """
+        normal_output = self.main_partner.lang
+        self.assertEqual(normal_output, "en_US")
+
+        alt_output = self.main_partner.selection_value_displayname("lang")
+        self.assertEqual(alt_output, "English (US)")
+
+    def test_generated_selection_field2(self):
+        """Check function gives the expected value from a selection field
+        which uses a function to determine its selection values
+        (where the function is defined in the field as a string)
+        """
+        normal_output = self.ir_filters_recordset.model_id
+        self.assertEqual(normal_output, "res.groups")
+
+        alt_output = self.ir_filters_recordset.selection_value_displayname("model_id")
+        self.assertEqual(alt_output, "Access Groups")
+
+    def test_generated_lambda_selection_field(self):
+        """Check function gives the expected value from a selection field
+        which uses a lambda function to determine its selection values
+        """
+        normal_output = self.res_partner_bank_recordset.acc_type
+        self.assertEqual(normal_output, "bank")
+
+        alt_output = self.res_partner_bank_recordset.selection_value_displayname("acc_type")
+        self.assertEqual(alt_output, "Normal")
+
+    def test_fetching_nonexistent_field(self):
+        """Ensure function fails with appropriate warning message if developer
+        attempts to get a selection field which does not exist on the model
+        in place of a KeyError
+        """
+        with self.assertRaises(ValidationError):
+            self.res_partner_bank_recordset.selection_value_displayname("pin_number")
+
+    def test_fetching_unset_selection_field(self):
+        """Ensure function returns False if the selection field on the recordset is not set"""
+        self.model_overview_report.report_type = False
+
+        normal_output = self.model_overview_report.report_type
+        self.assertEqual(normal_output, False)
+
+        alt_output = self.model_overview_report.selection_value_displayname("report_type")
+        self.assertEqual(alt_output, False)
+
+    # NB: It's not possible (as far as i'm aware) to mock a new field on a model without lots of pain
```

Comment:
Left this in incase anyone has ideas

Reply:
Removed this as there is really no need to test it as we know it works and it's too tricky to mock